### PR TITLE
New add card workflow

### PR DESCRIPTION
**Pull request by TheLastProject**
Should improve #1511

Adding card without barcode:

| Before | After |
| - | - |
| image | image |
| image | image |
| | image |

Manually adding card with barcode:
| Before | After |
| - | - |
| image | image |
| image | image |

### CODE REVIEW

**Comment by obfusk** [comment]
I feel like the icon is wrong for that button. But otherwise looks nice! Will test & review if I can find the time.

**Comment by TheLastProject** [comment]
> I feel like the icon is wrong for that button

I mean, I didn't look into it super long but it kinda makes sense to me because it "expands" into a list. I'm open to suggestions but I do think we want to prevent using a symbol like `+` as that on a quick glance might easily be mistaken for "Add card"

**Comment by obfusk** [comment]
Yeah, I just feel like that particular icon on that particular UI element looks like the UI element itself should expand and not open a dialog. So it feels weird. A `+` would be bad yeah. Maybe a `⋮` works?

**Comment by TheLastProject** [comment]
Given your argument is "the icon implies it opens right there but it opens a dialog", `⋮` is not a good replacement given that icon is literally used by Android itself to open more menu options at the spot you clicked it.

**Comment by obfusk** [comment]
Somehow that icon looks less wrong to me. But yeah, good point. Will try to think of something else.

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -86,7 +86,7 @@
     <string name="app_libraries">Libre third-party libraries: <xliff:g id="app_libraries_list">%s</xliff:g></string>
     <string name="app_resources">Libre third-party resources: <xliff:g id="app_resources_list">%s</xliff:g></string>
     <string name="selectBarcodeTitle">Select Barcode</string>
-    <string name="enterBarcodeInstructions">Enter the ID, and either pick a barcode type below, or “There is no barcode”.</string>
+    <string name="enterBarcodeInstructions">Enter the ID number or text on your card and press the barcode that looks like the one on your card.</string>
```
Comment (obfusk) [inline app/src/main/res/values/strings.xml:89]: Outdated translations might still continue to refer to "There is no barcode" which no longer exists. Seems potentially confusing. But replacing the string with another key and thus have no translation at all doesn't seem much better either. Just pointing it out.

File: app/src/main/res/layout/barcode_selector_activity.xml [hunk 2]
```diff
@@ -65,16 +65,6 @@
...
         </LinearLayout>
-        <Button
-            android:id="@+id/noBarcode"
-            android:layout_width="match_parent"
-            android:layout_height="wrap_content"
-            app:layout_constraintTop_toBottomOf="@+id/barcodeIdLayout"
-            app:layout_constraintBottom_toTopOf="@+id/barcodes"
-            app:layout_constraintEnd_toEndOf="parent"
-            app:layout_constraintStart_toStartOf="parent"
-            android:text="@string/barcodeNoBarcode"
```
Comment (obfusk) [inline app/src/main/res/layout/barcode_selector_activity.xml:76]: This leaves the string `barcodeNoBarcode` unused.

File: app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -332,4 +332,9 @@
     <string name="show_archived_cards">Show archived cards</string>
     <string name="view_online">View online</string>
     <string name="field_may_not_be_empty">Field may not be empty</string>
+    <string name="action_more_options">More options</string>
+    <string name="addWithoutBarcode">Add a card without barcode</string>
+    <string name="enter_card_id">Enter the ID number or text on your card</string>
+    <string name="card_id_may_not_be_empty">Card ID may not be empty</string>
```
Comment (obfusk) [inline app/src/main/res/values/strings.xml:338]: I dislike "may not be"; "should not be" or "must not be" is less ambiguous, which is also easier for translators.

Also applies to `field_may_not_be_empty` I guess.

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -332,4 +332,9 @@
...
+    <string name="add_a_card_in_a_different_way">Add a card in a different way</string>
```
Comment (obfusk) [inline app/src/main/res/values/strings.xml:339]: The wording "in a different way" feels a bit off. But not sure what's better.

File: app/src/main/java/protect/card_locker/ScanActivity.java [hunk 5]
```diff
@@ -65,6 +76,10 @@ public class ScanActivity extends CatimaAppCompatActivity {
     // can't use the pre-made contract because that launches the file manager for image type instead of gallery
     private ActivityResultLauncher<Intent> photoPickerLauncher;
 
+    static final String STATE_SCANNER_ACTIVE = "scannerActive";
+    private boolean mScannerActive = true;
+
+
```
Comment (obfusk) [inline app/src/main/java/protect/card_locker/ScanActivity.java:82]: Is two blank lines intentional?

File: app/src/main/java/protect/card_locker/ScanActivity.java [hunk 6]
```diff
@@ -190,19 +246,21 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
-        if (barcodeValues.isEmpty()) {
-            return;
+    private void setScannerActive(boolean isActive) {
+        if (isActive) {
+            barcodeScannerView.resume();
+            mScannerActive = true;
+        } else {
+            barcodeScannerView.pause();
+            mScannerActive = false;
         }
```
Comment (obfusk) [inline app/src/main/java/protect/card_locker/ScanActivity.java:256]: Feels slightly cleaner. But I don't feel very strongly about it.

```suggestion
        if (isActive) {
            barcodeScannerView.resume();
        } else {
            barcodeScannerView.pause();
        }
        mScannerActive = isActive;
```

File: app/src/main/res/values/strings.xml [hunk 7]
```diff
@@ -157,8 +157,8 @@
     <string name="moveDown">Move downwards</string>
     <string name="leaveWithoutSaveTitle">Exit</string>
     <string name="leaveWithoutSaveConfirmation">Leave without saving?</string>
-    <string name="addManually">Manually enter ID</string>
-    <string name="addFromImage">Select image from gallery</string>
+    <string name="addManually">Select the barcode manually</string>
```
Comment (obfusk) [inline app/src/main/res/values/strings.xml:160]: To me, "select the barcode manually" sounds like it's missing the "enter the text/ID" part.

```suggestion
    <string name="addManually">Input the barcode manually</string>
```

File: app/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -332,4 +332,9 @@
     <string name="show_archived_cards">Show archived cards</string>
     <string name="view_online">View online</string>
     <string name="field_may_not_be_empty">Field may not be empty</string>
+    <string name="action_more_options">More options</string>
+    <string name="addWithoutBarcode">Add a card without barcode</string>
```
Comment (obfusk) [inline app/src/main/res/values/strings.xml:336]:
```suggestion
    <string name="addWithoutBarcode">Add a card without a barcode</string>
```

File: app/src/main/res/layout/custom_barcode_scanner.xml [hunk 9]
```diff
@@ -2,6 +2,15 @@
 <merge xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:app="http://schemas.android.com/apk/res-auto">
 
+    <com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton
+        android:id="@+id/fabOtherOptions"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_gravity="end|bottom"
+        app:icon="@drawable/ic_baseline_unfold_more_24"
```
Comment (obfusk) [inline app/src/main/res/layout/custom_barcode_scanner.xml:10]: I think it looks better if you just remove the icon. Unless you can find something that really fits. idk.

```suggestion
```

**Comment by obfusk** [comment]
Pressing the back button after choosing addManually or addFromImage will not restart the scan activity; fix (tested on my phone):

```diff
--- a/app/src/main/java/protect/card_locker/ScanActivity.java
+++ b/app/src/main/java/protect/card_locker/ScanActivity.java
@@ -272,6 +272,11 @@ public class ScanActivity extends CatimaAppCompatActivity {
     private void handleActivityResult(int requestCode, int resultCode, Intent intent) {
         super.onActivityResult(requestCode, resultCode, intent);

+        if (resultCode == Activity.RESULT_CANCELED) {
+            setScannerActive(true);
+            return;
+        }
+
         BarcodeValues barcodeValues = Utils.parseSetBarcodeActivityResult(requestCode, resultCode, intent, this);

         if (barcodeValues.isEmpty()) {
```

**Comment by obfusk** [comment]
LGTM otherwise.

**Comment by obfusk** [comment]
I'm not sure how the `PERMISSION_SCAN_ADD_FROM_IMAGE` stuff works; AFAIK it only applies on android < N.
So there might be another bug where `setScannerActive(true)` is not called when `onMockedRequestPermissionsResult()` never calls `addFromImageAfterPermission()` if the permission isn't granted.
Should test that on the older android versions (< N) that need the permission.

**Comment by obfusk** [comment]
And `photoPickerLauncher.launch()` in `addFromImageAfterPermission()` could also fail, in which case `setScannerActive(true)` isn't called either.

File: app/src/main/res/layout/custom_barcode_scanner.xml (hunk 9 above)
Comment (TheLastProject) [reply app/src/main/res/layout/custom_barcode_scanner.xml:10]: It doesn't really feel better to me, the button feels empty to me. But this is very personal taste, I don't think there's an objective better way.

**Comment by TheLastProject** [comment]
There were actually 2 bugs in `handleActivityResult` (you only noticed one of them). The second is when you choose a picture that has no barcode :sweat_smile: 

But I guess we do need to test Android < N, yeah.

File: app/src/main/res/layout/barcode_selector_activity.xml [hunk 10]
```diff
@@ -34,15 +34,15 @@
...
+            android:text="@string/manually_enter_barcode_instructions" />
         <LinearLayout
             android:id="@+id/barcodeIdLayout"
             android:orientation="horizontal"
             android:padding="10.0dip"
             android:layout_width="fill_parent"
             android:layout_height="wrap_content"
             app:layout_constraintTop_toBottomOf="@+id/explanationText"
-            app:layout_constraintBottom_toTopOf="@+id/noBarcode"
+            app:layout_constraintBottom_toTopOf="@+id/barcodes"
```
Comment (obfusk) [inline app/src/main/res/layout/barcode_selector_activity.xml:45]: I missed those. Nice catch!

File: app/src/main/res/layout/custom_barcode_scanner.xml (hunk 9 above)
Comment (obfusk) [reply app/src/main/res/layout/custom_barcode_scanner.xml:10]: Yeah. To me the current icon suggest the UI element itself will expand. So it feels off when it opens a dialog instead. But I get your PoV too. I wanted to give you my PoV. And I'll let you know if I ever think of a more appropriate icon. But it's up to you to use whatever you feel is best.

**Comment by obfusk** [comment]
> There were actually 2 bugs in `handleActivityResult` (you only noticed one of them).

Nice catch! Told you edge cases are easy to miss ;p

File: app/src/main/java/protect/card_locker/ScanActivity.java [hunk 11]
```diff
@@ -272,9 +269,15 @@ private void returnResult(String barcodeContents, String barcodeFormat) {
     private void handleActivityResult(int requestCode, int resultCode, Intent intent) {
         super.onActivityResult(requestCode, resultCode, intent);
 
+        if (resultCode == Activity.RESULT_CANCELED) {
```
Comment (obfusk) [inline app/src/main/java/protect/card_locker/ScanActivity.java:272]: With the second fix below, this is technically superfluous as `parseSetBarcodeActivityResult()` will return an empty `BarcodeValues` if `resultCode == Activity.RESULT_CANCELED`. But keeping it to be more explicit about the control flow on cancel also has benefits IMO. Up to you.

**Comment by obfusk** [comment]
My guess is we need this for android < N (needs testing):

```diff
diff --git a/app/src/main/java/protect/card_locker/ScanActivity.java b/app/src/main/java/protect/card_locker/ScanActivity.java
index 17597669..7138937b 100644
--- a/app/src/main/java/protect/card_locker/ScanActivity.java
+++ b/app/src/main/java/protect/card_locker/ScanActivity.java
@@ -373,6 +373,7 @@ public class ScanActivity extends CatimaAppCompatActivity {
         try {
             photoPickerLauncher.launch(chooserIntent);
         } catch (ActivityNotFoundException e) {
+            setScannerActive(true);
             Toast.makeText(getApplicationContext(), R.string.failedLaunchingPhotoPicker, Toast.LENGTH_LONG).show();
             Log.e(TAG, "No activity found to handle intent", e);
         }
@@ -425,6 +426,7 @@ public class ScanActivity extends CatimaAppCompatActivity {
             if (granted) {
                 addFromImageAfterPermission();
             } else {
+                setScannerActive(true);
                 Toast.makeText(this, R.string.storageReadPermissionRequired, Toast.LENGTH_LONG).show();
             }
         }
```

File: app/src/main/java/protect/card_locker/ScanActivity.java (hunk 11 above)
Comment (obfusk) [reply app/src/main/java/protect/card_locker/ScanActivity.java:272]: > technically superfluous

Because `parseSetBarcodeActivityResult()` starts with:

```java
        if (resultCode != Activity.RESULT_OK) {
            return new BarcodeValues(null, null);
        }
```

I just overlooked that it would also be empty if it failed to get a barcode from the image. Was mostly looking at ways the flow could be cancelled in a way that would fail to set the scanner to active again.

**Comment by obfusk** [comment]
I'll do some more on-device testing for edge cases before the next release after this gets merged; same with #1517.

**Comment by TheLastProject** [comment]
> My guess is we need this for android < N (needs testing)

I've been spending the past hour trying to get Android Studio to give me a camera preview so I can test but... nope. So this is... becoming very hard to test.

**Comment by obfusk** [comment]
I tested this patch on Android Marshmallow in the emulator. Everything works as expected. Without the patch it failed to resume the scan if the permission was denied.

```diff
diff --git a/app/src/main/java/protect/card_locker/ScanActivity.java b/app/src/main/java/protect/card_locker/ScanActivity.java
index 17597669..8a0b95c3 100644
--- a/app/src/main/java/protect/card_locker/ScanActivity.java
+++ b/app/src/main/java/protect/card_locker/ScanActivity.java
@@ -269,11 +269,6 @@ public class ScanActivity extends CatimaAppCompatActivity {
     private void handleActivityResult(int requestCode, int resultCode, Intent intent) {
         super.onActivityResult(requestCode, resultCode, intent);
 
-        if (resultCode == Activity.RESULT_CANCELED) {
-            setScannerActive(true);
-            return;
-        }
-
         BarcodeValues barcodeValues = Utils.parseSetBarcodeActivityResult(requestCode, resultCode, intent, this);
 
         if (barcodeValues.isEmpty()) {
@@ -373,6 +368,7 @@ public class ScanActivity extends CatimaAppCompatActivity {
         try {
             photoPickerLauncher.launch(chooserIntent);
         } catch (ActivityNotFoundException e) {
+            setScannerActive(true);
             Toast.makeText(getApplicationContext(), R.string.failedLaunchingPhotoPicker, Toast.LENGTH_LONG).show();
             Log.e(TAG, "No activity found to handle intent", e);
         }
@@ -425,6 +421,7 @@ public class ScanActivity extends CatimaAppCompatActivity {
             if (granted) {
                 addFromImageAfterPermission();
             } else {
+                setScannerActive(true);
                 Toast.makeText(this, R.string.storageReadPermissionRequired, Toast.LENGTH_LONG).show();
             }
         }
```
